packages/ses/src/module-load.js

```diff
@@ -145,83 +139,72 @@ function* loadWithoutErrorAnnotation(
+
+  if (typeof moduleDescriptor === 'string') {
     // eslint-disable-next-line @endo/no-polymorphic-call
     assert.fail(
       X`Cannot map module ${q(moduleSpecifier)} to ${q(
-        aliasNamespace,
+        moduleDescriptor,
       )} in parent compartment, not yet implemented`,
       TypeError,
     );
-  } else if (aliasNamespace !== undefined) {
-    const alias = weakmapGet(moduleAliases, aliasNamespace);
-    if (alias === undefined) {
-      // eslint-disable-next-line @endo/no-polymorphic-call
-      assert.fail(
-        X`Cannot map module ${q(
-          moduleSpecifier,
-        )} because the value is not a module exports namespace, or is from another realm`,
-        ReferenceError,
-      );
+  } else if (isObject(moduleDescriptor)) {
+    // In this shim (and not in XS, and not in the standard we imagine), we
+    // allow a module namespace object to stand in for a module descriptor that
+    // describes its original {compartment, specifier} so that it can be used
+    // to create a link.
+    const aliasDescriptor = weakmapGet(moduleAliases, moduleDescriptor);
+    if (aliasDescriptor !== undefined) {
+      moduleDescriptor = aliasDescriptor;
     }
-    // Behold: recursion.
-    // eslint-disable-next-line no-use-before-define
-    const aliasRecord = yield memoizedLoadWithErrorAnnotation(
-      compartmentPrivateFields,
-      moduleAliases,
-      alias.compartment,
-      alias.specifier,
-      enqueueJob,
-      selectImplementation,
-      moduleLoads,
-    );
-    mapSet(moduleRecords, moduleSpecifier, aliasRecord);
-    return aliasRecord;
-  }
-
-  if (mapHas(moduleRecords, moduleSpecifier)) {
-    return mapGet(moduleRecords, moduleSpecifier);
-  }
-
-  const staticModuleRecord = yield selectImplementation(
-    importHook,
-    importNowHook,
-  )(moduleSpecifier);
-
-  if (staticModuleRecord === null || typeof staticModuleRecord !== 'object') {
-    Fail`importHook must return a promise for an object, for module ${q(
-      moduleSpecifier,
-    )} in compartment ${q(compartment.name)}`;
-  }
 
-  // check if record is a RedirectStaticModuleInterface
-  if (staticModuleRecord.specifier !== undefined) {
-    // check if this redirect with an explicit record
-    if (staticModuleRecord.record !== undefined) {
-      // ensure expected record shape
-      if (staticModuleRecord.compartment !== undefined) {
-        throw TypeError(
-          'Cannot redirect to an explicit record with a specified compartment',
-        );
-      }
+    // A (legacy) module descriptor for when we find the module source (record)
+    // but at a different specifier than requested.
+    // Providing this {specifier, record} descriptor serves as an ergonomic
+    // short-hand for stashing the record, returning a {compartment, specifier}
+    // reference, bouncing the module hook, then producing the source (record)
+    // when module hook receives the response specifier.
+    if (moduleDescriptor.record !== undefined) {
```

Comment:
Descriptors with the `record` property are spiritually equivalent to the XS `source` module descriptor, to be introduced in the subsequent PR. The only behavior `record` descriptors support is the creation of a module that has a different “response specifier” than the “request specifier”. That is, `importHook(requestSpecifier)` returns `{ record: moduleSource, specifier: responseSpecifier }`, such that all the import specifiers in the source are resolved relative to responseSpecifier (e.g., `./src/index.js`) instead of the requestSpecifier (e.g., `.`).

XS `source` module descriptors will support more shapes and treat the `compartment` default differently. The `source` may be a string instead of a module source, in which case it’s a specifier to load the source from another compartment. The default compartment is the parent instead of the child.